File: low-level-design/lld-coding/multi-language/04-page-view-counter/PageViewCounter.java
/*
 * Java: Page View Counter - Race Condition Demo
 * ==============================================
 * Shows:
 * 1. Broken version (race condition)
 * 2. Fixed version (synchronized keyword)
 * 3. Fixed version (ReentrantLock)
 * 4. Fixed version (AtomicInteger)
 */

import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class PageViewCounter {

    public static void main(String[] args) throws InterruptedException {
        System.out.println("================================================================================");
        System.out.println("          JAVA: RACE CONDITION DEMONSTRATION");
        System.out.println("================================================================================");

        // Version 1: Broken
        demonstrateBroken();

        // Version 2: Fixed with synchronized
        demonstrateSynchronized();

        // Version 3: Fixed with ReentrantLock
        demonstrateReentrantLock();

        // Version 4: Fixed with AtomicInteger
        demonstrateAtomic();

        // Summary
        printSummary();
    }

    // ============================================================================
    // Version 1: BROKEN - Race Condition
    // ============================================================================
    static class PageViewCounterBroken {
        private int views = 0;

        public void increment() {
            // This looks simple but is NOT atomic!
            int temp = views;
            try {
                Thread.sleep(0, 100000); // 0.1ms - increase race window
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            views = temp + 1;
        }

        public int getViews() {
            return views;
        }
    }

    static void demonstrateBroken() throws InterruptedException {
        System.out.println("\n1. BROKEN VERSION (No Synchronization)");
        System.out.println("--------------------------------------------------------------------------------");

        PageViewCounterBroken counter = new PageViewCounterBroken();
        Thread[] threads = new Thread[1000];

        // Create 1000 threads
        for (int i = 0; i < 1000; i++) {
            threads[i] = new Thread(() -> counter.increment());
            threads[i].start();
        }

        // Wait for all threads
        for (Thread t : threads) {
            t.join();
        }

        System.out.println("Expected: 1000");
        System.out.println("Actual:   " + counter.getViews());
        System.out.println("Lost:     " + (1000 - counter.getViews()) + " increments");
        System.out.println("Status:   " + (counter.getViews() == 1000 ? "✅ Correct" : "❌ RACE CONDITION DETECTED!"));
    }

    // ============================================================================
    // Version 2: FIXED - Using synchronized
    // ============================================================================
    static class PageViewCounterSynchronized {
        private int views = 0;

        // synchronized keyword makes method thread-safe
        public synchronized void increment() {
            int temp = views;
            try {
                Thread.sleep(0, 100000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            views = temp + 1;
        }

        public synchronized int getViews() {
            return views;
        }
    }

    static void demonstrateSynchronized() throws InterruptedException {
        System.out.println("\n2. FIXED VERSION (With synchronized)");
        System.out.println("--------------------------------------------------------------------------------");

        PageViewCounterSynchronized counter = new PageViewCounterSynchronized();
        Thread[] threads = new Thread[1000];

        for (int i = 0; i < 1000; i++) {
            threads[i] = new Thread(() -> counter.increment());
            threads[i].start();
        }

        for (Thread t : threads) {
            t.join();
        }

        System.out.println("Expected: 1000");
        System.out.println("Actual:   " + counter.getViews());
        System.out.println("Status:   " + (counter.getViews() == 1000 ? "✅ Correct!" : "❌ Still broken?!"));
    }

    // ============================================================================
    // Version 3: FIXED - Using ReentrantLock
    // ============================================================================
    static class PageViewCounterReentrantLock {
        private int views = 0;
        private final ReentrantLock lock = new ReentrantLock();

        public void increment() {
            lock.lock(); // Acquire lock
            try {
                int temp = views;
                Thread.sleep(0, 100000);
                views = temp + 1;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                lock.unlock(); // Always unlock in finally block
            }
        }

        public int getViews() {
            lock.lock();
            try {
                return views;
            } finally {
                lock.unlock();
            }
        }
    }

    static void demonstrateReentrantLock() throws InterruptedException {
        System.out.println("\n3. FIXED VERSION (With ReentrantLock)");
        System.out.println("--------------------------------------------------------------------------------");
        System.out.println("More flexible than synchronized, supports try-lock, fairness, etc.");

        PageViewCounterReentrantLock counter = new PageViewCounterReentrantLock();
        Thread[] threads = new Thread[1000];

        for (int i = 0; i < 1000; i++) {
            threads[i] = new Thread(() -> counter.increment());
            threads[i].start();
        }

        for (Thread t : threads) {
            t.join();
        }

        System.out.println("Expected: 1000");
        System.out.println("Actual:   " + counter.getViews());
        System.out.println("Status:   ✅ Correct!");
    }

    // ============================================================================
    // Version 4: BEST PRACTICE - Using AtomicInteger
    // ============================================================================
    static class PageViewCounterAtomic {
        private final AtomicInteger views = new AtomicInteger(0);

        public void increment() {
            // Lock-free, thread-safe increment
            views.incrementAndGet();
        }

        public int getViews() {
            return views.get();
        }
    }

    static void demonstrateAtomic() throws InterruptedException {
        System.out.println("\n4. BEST PRACTICE (AtomicInteger)");
        System.out.println("--------------------------------------------------------------------------------");
        System.out.println("For simple counters, AtomicInteger is lock-free and faster");

        PageViewCounterAtomic counter = new PageViewCounterAtomic();

        // Use ExecutorService for better thread management
        ExecutorService executor = Executors.newFixedThreadPool(100);

        // Simulate 100 users, each viewing 10 pages
        for (int i = 0; i < 100; i++) {
            executor.submit(() -> {
                for (int j = 0; j < 10; j++) {
                    counter.increment();
                    try {
                        Thread.sleep(0, 10000); // 0.01ms
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }
            });
        }

        executor.shutdown();
        executor.awaitTermination(10, TimeUnit.SECONDS);

        System.out.println("Expected: 1000 (100 users × 10 views)");
        System.out.println("Actual:   " + counter.getViews());
        System.out.println("Status:   ✅ Production-ready!");
    }

    // ============================================================================
    // Summary
    // ============================================================================
    static void printSummary() {
        System.out.println("\n================================================================================");
        System.out.println("                          SUMMARY");
        System.out.println("================================================================================");
        System.out.print("""

🎯 Key Lessons:

1. Java has TRUE parallelism (no GIL)
   - Multiple threads run simultaneously
   - Race conditions are common
   - MUST use synchronization

2. Synchronization options in Java:
   a) synchronized keyword
      - Simplest, built into language
      - synchronized method or synchronized(obj) block
      - Implicit lock per object

   b) ReentrantLock (java.util.concurrent.locks)
      - More flexible than synchronized
      - Supports tryLock(), fairness, conditions
      - Must unlock in finally block

   c) Atomic classes (java.util.concurrent.atomic)
      - Lock-free, fastest
      - AtomicInteger, AtomicLong, AtomicReference
      - Only for simple operations

3. When to use what:
   - Simple counter? → AtomicInteger
   - Complex logic? → synchronized or ReentrantLock
   - Need tryLock? → ReentrantLock
   - Default choice? → synchronized (simplest)

4. Best practices:
   - Prefer atomic classes for counters
   - Use ExecutorService over raw threads
   - Always unlock in finally block (ReentrantLock)
   - Keep synchronized blocks small
   - Consider java.util.concurrent collections

5. Thread pools:
   - Don't create raw threads in production
   - Use ExecutorService
   - Fixed, cached, scheduled thread pools

⚠️  Java makes concurrency explicit - use it correctly or suffer races!
""");
    }
}

File: low-level-design/lld-coding/multi-language/04-page-view-counter/page_view_counter.go
/*
Go: Page View Counter - Race Condition Demo
============================================
Shows:
1. Broken version (race condition)
2. Fixed version (with Mutex)
3. Go's race detector
4. Alternative: atomic operations
*/

package main

import (
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

func main() {
	fmt.Println("================================================================================")
	fmt.Println("          GO: RACE CONDITION DEMONSTRATION")
	fmt.Println("================================================================================")

	// Version 1: Broken
	demonstrateBroken()

	// Version 2: Fixed with Mutex
	demonstrateFixed()

	// Version 3: Fixed with Atomic
	demonstrateAtomic()

	// Summary
	printSummary()
}

// ============================================================================
// Version 1: BROKEN - Race Condition
// ============================================================================
type PageViewCounterBroken struct {
	views int
}

func (c *PageViewCounterBroken) Increment() {
	// This is NOT atomic!
	temp := c.views
	time.Sleep(100 * time.Microsecond) // Increase race window
	c.views = temp + 1
}

func demonstrateBroken() {
	fmt.Println("\n1. BROKEN VERSION (No Mutex)")
	fmt.Println("--------------------------------------------------------------------------------")

	counter := &PageViewCounterBroken{views: 0}
	var wg sync.WaitGroup

	// Create 1000 goroutines (much lighter than threads!)
	for i := 0; i < 1000; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			counter.Increment()
		}()
	}

	// Wait for all goroutines to complete
	wg.Wait()

	fmt.Printf("Expected: 1000\n")
	fmt.Printf("Actual:   %d\n", counter.views)
	fmt.Printf("Lost:     %d increments\n", 1000-counter.views)
	if counter.views == 1000 {
		fmt.Println("Status:   ✅ Correct (got lucky!)")
	} else {
		fmt.Println("Status:   ❌ RACE CONDITION DETECTED!")
	}
	fmt.Println("\n💡 Run with: go run -race page_view_counter.go")
	fmt.Println("   Go's race detector will catch this!")
}

// ============================================================================
// Version 2: FIXED - Using Mutex
// ============================================================================
type PageViewCounterFixed struct {
	views int
	mu    sync.Mutex // Mutex for synchronization
}

func (c *PageViewCounterFixed) Increment() {
	c.mu.Lock() // Acquire lock
	defer c.mu.Unlock() // Release lock when function returns

	temp := c.views
	time.Sleep(100 * time.Microsecond)
	c.views = temp + 1
}

func (c *PageViewCounterFixed) GetViews() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.views
}

func demonstrateFixed() {
	fmt.Println("\n2. FIXED VERSION (With Mutex)")
	fmt.Println("--------------------------------------------------------------------------------")

	counter := &PageViewCounterFixed{}
	var wg sync.WaitGroup

	for i := 0; i < 1000; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			counter.Increment()
		}()
	}

	wg.Wait()

	fmt.Printf("Expected: 1000\n")
	fmt.Printf("Actual:   %d\n", counter.GetViews())
	if counter.GetViews() == 1000 {
		fmt.Println("Status:   ✅ Correct!")
	} else {
		fmt.Println("Status:   ❌ Still broken?!")
	}
}

// ============================================================================
// Version 3: BEST PRACTICE - Using Atomic Operations
// ============================================================================
type PageViewCounterAtomic struct {
	views int64 // Must be int64 for atomic operations
}

func (c *PageViewCounterAtomic) Increment() {
	// atomic.AddInt64 is lock-free and faster than mutex
	atomic.AddInt64(&c.views, 1)
}

func (c *PageViewCounterAtomic) GetViews() int64 {
	return atomic.LoadInt64(&c.views)
}

func demonstrateAtomic() {
	fmt.Println("\n3. BEST PRACTICE (Atomic Operations)")
	fmt.Println("--------------------------------------------------------------------------------")
	fmt.Println("For simple counters, atomic operations are faster than mutex")

	counter := &PageViewCounterAtomic{}
	var wg sync.WaitGroup

	// Simulate 100 users, each viewing 10 pages
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			// Each goroutine increments 10 times
			for j := 0; j < 10; j++ {
				counter.Increment()
				time.Sleep(10 * time.Microsecond)
			}
		}()
	}

	wg.Wait()

	fmt.Printf("Expected: 1000 (100 users × 10 views)\n")
	fmt.Printf("Actual:   %d\n", counter.GetViews())
	fmt.Println("Status:   ✅ Production-ready and lock-free!")
}

// ============================================================================
// Summary
// ============================================================================
func printSummary() {
	fmt.Println("\n================================================================================")
	fmt.Println("                          SUMMARY")
	fmt.Println("================================================================================")
	fmt.Print(`
🎯 Key Lessons:

1. Go has TRUE parallelism (no GIL like Python)
   - Multiple goroutines run simultaneously on multiple cores
   - Race conditions are very common
   - ALWAYS use synchronization

2. Goroutines are lightweight
   - Can create thousands easily
   - Much lighter than OS threads
   - Managed by Go runtime

3. Synchronization options:
   a) sync.Mutex - General purpose locking
      - Use for complex operations
      - Explicit Lock()/Unlock()

   b) atomic package - For simple operations
      - Lock-free, faster
      - Only for basic types (int32, int64, pointers)

   c) Channels - For message passing
      - "Don't communicate by sharing memory; share memory by communicating"
      - Idiomatic Go way

4. Go race detector is your friend!
   - go run -race program.go
   - go test -race
   - Catches races at runtime

5. Best practices:
   - Embed sync.Mutex in structs
   - Use defer to unlock
   - Prefer atomic for counters
   - Use channels for coordination

⚠️  Go makes concurrency easy - but race conditions are still dangerous!
    Always test with -race flag!
`)
}
